ipn/ipnlocal/profiles.go

```diff
@@ -231,37 +233,69 @@ func (pm *profileManager) SetPrefs(prefsIn ipn.PrefsView, np ipn.NetworkProfile)
 				// We couldn't delete the state, so keep the profile around.
 				continue
 			}
-			// Remove the profile, knownProfiles will be persisted below.
+			// Remove the profile, knownProfiles will be persisted
+			// in [profileManager.setProfilePrefs] below.
 			delete(pm.knownProfiles, p.ID)
 		}
-	} else if cp.ID == "" {
-		// We didn't have an existing profile, so create a new one.
-		cp.ID, cp.Key = newUnusedID(pm.knownProfiles)
-		cp.LocalUserID = pm.currentUserID
-	} else {
-		// This means that there was a force-reauth as a new node that
-		// we haven't seen before.
-	}
-
-	if prefs.ProfileName != "" {
-		cp.Name = prefs.ProfileName
-	} else {
-		cp.Name = up.LoginName
 	}
-	cp.ControlURL = prefs.ControlURL
-	cp.UserProfile = newPersist.UserProfile
-	cp.NodeID = newPersist.NodeID
-	cp.NetworkProfile = np
-	pm.knownProfiles[cp.ID] = cp
 	pm.currentProfile = cp
-	if err := pm.writeKnownProfiles(); err != nil {
+	if err := pm.SetProfilePrefs(cp, prefsIn, np); err != nil {
 		return err
 	}
-	if err := pm.setAsUserSelectedProfileLocked(); err != nil {
+	return pm.setProfileAsUserDefault(cp)
+
+}
+
+// SetProfilePrefs is like [profileManager.SetPrefs], but sets prefs for the specified profile
+// which is not necessarily the current one. It returns an [errProfileAccessDenied] if the specified
+// profile is not accessible by the current user.
+func (pm *profileManager) SetProfilePrefs(profile *ipn.LoginProfile, prefsIn ipn.PrefsView, np ipn.NetworkProfile) error {
+	if err := pm.checkProfileAccess(profile); err != nil {
 		return err
 	}
-	if err := pm.setPrefsLocked(prefs.View()); err != nil {
-		return err
+
+	if profile.ID == "" {
+		if persist := prefsIn.Persist(); persist.Valid() && persist.NodeID() != "" && persist.UserProfile().LoginName != "" {
+			// We didn't have an existing profile, so create a new one.
+			profile.ID, profile.Key = newUnusedID(pm.knownProfiles)
+		}
+
+		// Set the current user as the profile owner, unless the current user ID does
+		// not represent a specific user, or the profile is already owned by a different user.
+		// It is only relevant on Windows where we have a multi-user system.
+		if profile.LocalUserID == "" && pm.currentUserID != "" {
+			profile.LocalUserID = pm.currentUserID
+		}
+	}
+
+	var userProfile tailcfg.UserProfile
+	if persist := prefsIn.Persist(); persist.Valid() {
+		userProfile = persist.UserProfile()
+		if userProfile.DisplayName == "" {
+			userProfile.DisplayName = userProfile.LoginName
+		}
+		profile.NodeID = persist.NodeID()
+	} else {
+		profile.NodeID = ""
+	}
+
+	if prefsIn.ProfileName() != "" {
+		profile.Name = prefsIn.ProfileName()
+	} else {
+		profile.Name = userProfile.LoginName
+	}
+	profile.ControlURL = prefsIn.ControlURL()
+	profile.UserProfile = userProfile
+	profile.NetworkProfile = np
```

Comment:
these renames are pretty distracting when trying to find the logic that has actually changed, can you pull these out into their own commit?

Reply:
or maybe keep `cp` for the current profile as its easier to reason about

Reply:
done. Since `SetProfilePrefs` accepts a `ipn.LoginProfile` as a parameter, and works with any profile, not just the current one, I kept the name `cp` in the first commit for your convenience but renamed it to `lp` in the second commit.